File: SimpleEngine/ArcBall.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine
{
    public class ArcBall
    {
        public ArcBall(vec2 center, float radius)
        {
            this.radius = radius;
            this.center = center;
        }

        public vec3 transform(vec2 v)
        {
            float x = (v.x - center.x) / radius;
            float y = (v.y - center.y) / radius;
            float rsq = x * x + y * y;
            float z = (rsq < 1) ? (float)(Math.Sqrt(1 - rsq)) : 0;
            return new vec3(x, y, z).Normalized;
        }

        public quat rotation(vec2 from, vec2 to)
        {
            vec3 f = transform(from);
            vec3 t = transform(to);
            vec3 axis = vec3.cross(f, t).Normalized;
            angle angle = vec3.AngleBetween(f, t);
            return new quat(AxisAngle.RightHandAround(axis, angle));
        }

        private readonly vec2 center;
        private readonly float radius;
    }
}

File: src/engine/arcball.cc
#include "engine/arcball.h"

#include <cmath>

#include "engine/axisangle.h"
#include "engine/quat.h"

namespace simple
{
    ArcBall::ArcBall(const vec2& c, float r)
        : center(c)
        , radius(r)
    {
    }

    vec3 Transform(const ArcBall& b, const vec2& v)
    {
        const auto x = (v.x - b.center.x) / b.radius;
        const auto y = (v.y - b.center.y) / b.radius;
        const auto rsq = x * x + y * y;
        const auto z = (rsq < 1) ? std::sqrt(1 - rsq) : 0.0f;
        return vec3(x, y, z).GetNormalized();
    }

    quat ArcBall::GetRotation(const vec2& from, const vec2& to) const
    {
        const auto f = Transform(*this, from);
        const auto t = Transform(*this, to);
        const auto axis = vec3::cross(f, t).GetNormalized();
        const auto angle = vec3::AngleBetween(f, t);
        return quat(AxisAngle::RightHandAround(axis, angle));
    }
}
